pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,299 @@
+	"sigs.k8s.io/kubefed/pkg/client/generic/scheme"
+)
+
+const (
+	tokenKey = "token"
 )
 
 func (f *federator) WatchClusters(handler federate.ClusterEventHandler) error {
-	panic("not implemented")
+	f.startInformerOnce.Do(func() {
+		f.kubeFedClusterInformer.AddEventHandler(f)
+
+		go func() {
+			f.kubeFedClusterInformer.Run(f.stopChan)
+			klog.Info("KubeFedCluster informer stopped")
+		}()
+
+		klog.Info("Started KubeFedCluster informer")
+	})
+
+	var handlerString string
+	if stringer, ok := handler.(fmt.Stringer); ok {
+		handlerString = stringer.String()
+	} else {
+		handlerString = reflect.TypeOf(handler).String()
+	}
+
+	clusterWatcher := &clusterWatcher{
+		eventQueue:           make(chan func(), 1000),
+		enqueueTimerInterval: time.Duration(time.Second * 10),
+		stopChan:             f.stopChan,
+		handler:              handler,
+		handlerString:        handlerString,
+	}
+
+	f.Lock()
+	defer f.Unlock()
+
+	f.clusterWatchers = append(f.clusterWatchers, clusterWatcher)
+
+	go clusterWatcher.eventLoop()
+
+	for clusterID, kubeConfig := range f.clusterMap {
+		clusterWatcher.onAdd(clusterID, kubeConfig)
+	}
+
+	klog.Infof("Federator added ClusterEventHandler \"%s\"", handlerString)
+
+	return nil
+}
+
+func (f *federator) OnAdd(obj interface{}) {
+	klog.V(3).Infof("In federated cluster watcher OnAdd for %#v", obj)
+
+	kubeFedCluster := obj.(*fedv1.KubeFedCluster)
+
+	kubeConfig, err := f.buildFederatedClusterConfig(kubeFedCluster)
+	if err != nil {
+		klog.Errorf("Error building kube config for federated cluster %#v: %v", kubeFedCluster, err)
+		return
+	}
+
+	f.Lock()
+	defer f.Unlock()
+
+	if f.clusterMap[kubeFedCluster.Name] != nil {
+		klog.V(2).Infof("An entry for KubeFedCluster \"%s\" already exists", kubeFedCluster.Name)
+		return
+	}
+
+	klog.Infof("KubeFedCluster \"%s\" has been added", kubeFedCluster.Name)
+
+	f.clusterMap[kubeFedCluster.Name] = kubeConfig
+
+	for _, clusterWatcher := range f.clusterWatchers {
+		clusterWatcher.onAdd(kubeFedCluster.Name, kubeConfig)
+	}
+}
+
+func (f *federator) OnDelete(obj interface{}) {
+	klog.V(3).Infof("In federated cluster watcher OnDelete for %#v", obj)
+
+	kubeFedCluster := obj.(*fedv1.KubeFedCluster)
+
+	f.Lock()
+	defer f.Unlock()
+
+	if f.clusterMap[kubeFedCluster.Name] == nil {
+		klog.Warningf("OnDelete - no cached entry for KubeFedCluster \"%s\" exists", kubeFedCluster.Name)
+		return
+	}
+
+	klog.Infof("KubeFedCluster \"%s\" has been deleted", kubeFedCluster.Name)
+
+	delete(f.clusterMap, kubeFedCluster.Name)
+
+	for _, clusterWatcher := range f.clusterWatchers {
+		clusterWatcher.onRemove(kubeFedCluster.Name)
+	}
+}
+
+func (f *federator) OnUpdate(oldObj, newObj interface{}) {
+	klog.V(3).Infof("In federated cluster watcher OnUpdate - OLD OBJ: %#v\nNEW OBJ: %#v", oldObj, newObj)
+
+	oldCluster := oldObj.(*fedv1.KubeFedCluster)
+	newCluster := newObj.(*fedv1.KubeFedCluster)
+
+	if reflect.DeepEqual(oldCluster.Spec, newCluster.Spec) {
```

Comment:
Quite a bit actually 😄 `KubeFedCluster` has other fields/structs like `KubeFedClusterStatus` that are updated periodically (eg `LastProbeTime`). I'm only interested in changes to the `KubeFedClusterSpec` hence the check here.

Reply:
ha, that gives me a hint on things I can ignore on my side.

May be a comment explaining why we deepcompare the spec would be helpful
to future code-readers :)

Reply:
Hmm I was thinking that at some point it could be interesting also to get more info on the cluster, like labels, etc... out of the Spec part. Not necessary now

Reply:
good to know and thus useful to document. Please add a comment :)

Reply:
will do